peps/pep-0810.rst

```diff
@@ -0,0 +1,1345 @@
+
+  The above exception was the direct cause of the following exception:
+
+  Traceback (most recent call last):
+    File "app.py", line 8, in <module>
+      result = dumsp({"key": "value"})
+               ^^^^^
+  ImportError: cannot import name 'dumsp' from 'json'. Did you mean: 'dump'?
+
+This exception chaining clearly shows: (1) where the lazy import was defined,
+(2) that it was deferred, and (3) where the actual access happened that triggered
+the error.
+
+Reification does **not** automatically occur when a module that was previously lazily
+imported is subsequently eagerly imported. Reification does **not** immediately
+resolve all lazy objects (e.g. ``lazy from`` statements) that referenced the module.
+It **only** resolves the lazy object being accessed.
+
+Accessing a lazy object (from a global variable or a module attribute)
+reifies the object. Accessing a module's ``__dict__`` reifies **all** lazy objects
+in that module. Operations that indirectly access ``__dict__`` (such as ``dir()``)
+also trigger this behavior.
+
+Example using ``__dict__`` from external code:
+
+.. code-block:: python
+
+  # my_module.py
+  import sys
+  lazy import json
+
+  print('json' in sys.modules)  # False - still lazy
+
+  # main.py
+  import sys
+  import my_module
+
+  # Accessing __dict__ from external code DOES reify all lazy imports
+  d = my_module.__dict__
+
+  print('json' in sys.modules)  # True - reified by __dict__ access
+  print(type(d['json']))  # <class 'module'>
+
+However, calling ``globals()`` does **not** trigger reification — it returns
+the module's dictionary, and accessing lazy objects through that dictionary
+still returns lazy proxy objects that need to be manually reified upon use.
+A lazy object can be resolved explicitly by calling the ``get`` method.
+Other, more indirect ways of accessing arbitrary globals (e.g. inspecting
+``frame.f_globals``) also do **not** reify all the objects.
+
+Example using ``globals()``:
+
+.. code-block:: python
+
+  import sys
+  lazy import json
+
+  # Calling globals() does NOT trigger reification
+  g = globals()
+
+  print('json' in sys.modules)  # False - still lazy
+  print(type(g['json']))  # <class 'lazy_import'>
+
+  # Explicitly reify using the get() method
+  resolved = g['json'].get()
+
+  print(type(resolved))  # <class 'module'>
+  print('json' in sys.modules)  # True - now loaded
+
+Bytecode and adaptive specialization
+-------------------------------------
+
+Lazy imports are implemented through modifications to four bytecode instructions:
+``IMPORT_NAME``, ``IMPORT_FROM``, ``LOAD_GLOBAL``, and ``LOAD_NAME``.
+
+The ``lazy`` syntax sets a flag in the ``IMPORT_NAME`` instruction's oparg
+(``oparg & 0x01``). The interpreter checks this flag and calls
+``_PyEval_LazyImportName()`` instead of ``_PyEval_ImportName()``, creating a lazy
+import object rather than executing the import immediately. The ``IMPORT_FROM``
```

Comment:
Why do import and import from use a different mechanism?

And how does this interact with the `__lazy_modules__` global? Is that syntactic (i.e. it affects the bytecode) or does it take effect at runtime?

Also, the discussion of bytecode and specific C API functions doesn't belong here: the Specification should specify the user-facing behavior, not implementation details in CPython.